In [28]:
import pandas as pd
import numpy as np


In [24]:
df = pd.read_csv("train.csv")

In [3]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

# 1. Popunim srednjom vrijednosti (int brojGodina, 24.3 -> float)
# 2. Obrisati recorde gdje nemamo tu vrijednosti - <5-10%
# 3. Popuni najblizom nenultom vrijednoscu JE DOBRO A I NIJE

In [ ]:
# datum temperatura
# 1.3.     5C
# 2.3.     6C
# 3.3.     7C
# 4.3.  
# 5.3.     
# 6.3.     12C

# 11.3.    15C
# 12.3.    16C
# 13.3.    15C
# 14.3.    14C
# 15.3.    17C
# 6.3.     12C


# ime visina
# ahmo. 199
# nudzejma 150
# mirzad  150





# 2 tipa podataka:

1. numericki (brojevni)
2. kategoricki (mozemo izabrati jedan ili vise ishoda od KONACNO MNOGO)

Kako popuniti ndeostajucu vrijednost ako su potencijalni ishodi : crvena, zuta i zelena?

In [ ]:
# pd.DataFrame()
# pd.Series()

In [5]:
type(df['Street'])

pandas.core.series.Series

In [10]:
type(df['LotFrontage'])

pandas.core.series.Series

In [6]:
type(df)

pandas.core.frame.DataFrame

In [7]:
df['Street'].describe()

count     1460
unique       2
top       Pave
freq      1454
Name: Street, dtype: object

In [9]:
df['LotFrontage'].describe()

count    1201.000000
mean       70.049958
std        24.284752
min        21.000000
25%        59.000000
50%        69.000000
75%        80.000000
max       313.000000
Name: LotFrontage, dtype: float64

In [8]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [17]:
df['Street'].value_counts().index[0]
# index: value -> lista
# key: value -> dict/object

'Pave'

In [21]:
df['Street'].value_counts().values[0]

np.int64(1454)

In [29]:
np.array([1,2,3])

array([1, 2, 3])

In [31]:
np.quantile([1,2,3], 0.5)
# YrSold

np.float64(2.0)

In [33]:
np.quantile(df['YrSold'], 0.5)

np.float64(2008.0)

In [61]:
# .fillna() 
# na -> not available

srednji_kvantil = np.nanquantile(
    df['LotFrontage'], 
    0.5
)

srednja_vrijednost = np.mean(df['LotFrontage'])

df['LotFrontage'] = df['LotFrontage'].fillna(
    srednja_vrijednost
)

# df['LotFrontage'] = df['LotFrontage'].fillna(0)

In [62]:
srednji_kvantil

np.float64(70.04995836802665)

In [ ]:
# 

In [55]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1460.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    22.024023    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    60.000000    7553.500000     5.000000   
50%     730.500000    50.000000    70.049958    9478.500000     6.000000   
75%    1095.250000    70.000000    79.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [56]:
# srednji_kvantil = np.quantile(
#     df['LotFrontage'], 
#     0.1
# )

uniqueValues = set()

for i in range(len(df['LotFrontage'])):
    uniqueValues.add(type(df['LotFrontage'].iloc[i]))


In [57]:
uniqueValues

{numpy.float64}

In [63]:
# spol
# Ž
# M 

In [ ]:
# SEMAFOR
# c 0
# ž 1
# z 2

In [ ]:
# one hot encoding

In [ ]:
# spol
# z
# z
# m
# z
# m



In [ ]:
# spol_z.  spol_m
#   1.       0
#   1.       0
#.   
#   1.       0
#.  0.       1

 1 dio dataseta mi je bilo greske/errors
 2 dio dataseta su bila mjeranja senzora

 cilj: sta ne valja na autu - predvidjeti

In [ ]:
# IQR
# inter quantile range

In [64]:
lf_srednja_vrijednost = np.mean(df['LotFrontage'])
lf_std = np.std(df['LotFrontage'])

In [65]:
print(lf_srednja_vrijednost)
print(lf_std)

70.04995836802664
22.0164789232215


In [67]:
donja_granica_IQR = lf_srednja_vrijednost - 3*lf_std
gornja_granica_IQR = lf_srednja_vrijednost + 3*lf_std

df = df[
    (df['LotFrontage'] > donja_granica_IQR) &
    (df['LotFrontage'] < gornja_granica_IQR) # |
]
# && 
# ||
# select * from df where LotFrontage > donja_granica_IQR and LotFrontage < gornja_granica_IQR


In [ ]:
# # DZ koji necu gledati
# # Odraditi isto ovo za sve numericke kolone jednom petljom
# for column in df.columns():
#     # Kako ispitati da li je column numeric
#     if column is numeric:
#         col_srednja_vrijednost = np.mean(df['LotFrontage'])
#         col_std = np.std(df['LotFrontage'])
#         donja_granica_IQR = col_srednja_vrijednost - 3*col_std
#         gornja_granica_IQR = col_srednja_vrijednost + 3*col_std
        
#         df = df[
#             (df[column] > donja_granica_IQR) &
#             (df[column] < gornja_granica_IQR) # |
#         ]

In [68]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1444.000000  1444.000000  1444.000000    1444.000000  1444.000000   
mean    729.774238    57.046399    68.895387   10261.858033     6.092105   
std     421.474086    42.456434    18.336553    8281.767930     1.376657   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     366.750000    20.000000    60.000000    7500.000000     5.000000   
50%     728.500000    50.000000    70.049958    9434.500000     6.000000   
75%    1092.250000    70.000000    78.000000   11475.750000     7.000000   
max    1460.000000   190.000000   134.000000  164660.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1444.000000  1444.000000   1444.000000  1436.000000  1444.000000  ...   
mean      5.578255  1971.273546   1984.806787   103.855850   437.945983  ...   
std       1.115057    30.213481     20.641599   180.572005   432.810181  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1993.000000     0.000000   381.000000  ...   
75%       6.000000  2000.000000   2004.000000   166.250000   709.000000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  2260.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1444.000000  1444.000000    1444.000000  1444.000000  1444.000000   
mean     94.442521    46.416898      21.675208     3.447368    15.227839   
std     125.553809    65.790302      59.624272    29.477207    56.042999   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     386.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1444.000000   1444.000000  1444.000000  1444.000000    1444.000000  
mean      1.718144     43.970914     6.326177  2007.811634  180118.227839  
std      32.894783    498.844699     2.703578     1.325888   77905.541601  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129900.000000  
50%       0.000000      0.000000     6.000000  2008.000000  162250.000000  
75%       0.000000      0.000000     8.000000  2009.000000  213310.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]